book/src/rfcs/ffadd.md

```diff
@@ -0,0 +1,192 @@
+
+
+```text
+bits  0..............87|88...........175|176...........263
+                       
+r  =  (-------r0-------|-------r1-------|-------r2-------)
++                      
+g  =  (-------g0-------|-------g1-------|-------g2-------)
+                       >                >               
+=                     k_0              k_1                
+
+u  =  (-------u0-------|-------u1-------|-------u2-------)
+```
+
+Following the steps above, and representing this equation in limb form, we have:
+
+\begin{eqnarray}
+u_0 &=& r_0 + g_0 - k_0 \cdot 2^{88} \\
+u_1 &=& r_1 + g_1 - k_1 \cdot 2^{88} + k_0\\
+u_2 &=& r_2 + g_2 + k_1\\
+\end{eqnarray}
+
+Finally, we perform a range check on the sum $u$, and we would know that $r < f$. 
+
+But now we also have to check that $0\leq r$ or, equivalently, $g\leq u$. Isn't the first check trivial because $r$ is a field element?
+
+
+### Subtractions
+
+Mathematically speaking, a subtraction within a field is no more than an addition over that field. Negative elements are not different from "positive" elements in finite fields (or in any modular arithmetic). To give a general example, the element $-e$ within a field $\mathbb{F}_m$ of order $m$ and $e<m$ is nothing but $m - e$. Nonetheless, for arbitrarily sized elements (not just those smaller than the modulus), the actual field element could be any $c \cdot m - e$, for any multiple $c \cdot m$ of the modulus. Thus, representing negative elements directly as "absolute" field elements may incur in additional computations involving multiplications and thus would result in a less efficient mechanism. 
+
+Alternatively, one can store negative values $-e$ as a negative sign followed by the absolute value $e$. Doing so will affect the way we address foreign field additions in the case that we had $a - b$ or $b - a$ (note that $- a - b$ is just the negation of $a + b$, and nothing changes at the gate level). Specifically, the carries between limbs may not just be bits ($0$ or $1$), but they could also be $-1$. This change only affects the values of $c_0, c_1, c_2$. 
+
+For the second part of the gate, meaning the sum for the range check, the carry bits will remain unchanged. The reason behind this is the fact that $g > 0$ because $2^{264} > f$, and $r$ will be "positive" as well. That means, this sum follows the usual positive-addition carry structure. 
+
+## Gadget
+
+The foreign field gadget will be composed by 4 sets of `RangeCheck` gadgets for witnesses $a, b, r, s$ accounting for 16 rows in total; followed by one row with `ForeignFieldAdd` gate type; and a final `Zero` row. A total of 18 rows with 15 columns in Kimchi.
```

Comment:
What about `u`?

Reply:
Oh, yeah here's a typo. By `s` I meant `u` there. I have to unify notation, definitely.